File: lua/b0RemoteApi.lua
require 'b0Lua'
b0.messagePack=require('messagePack-lua/MessagePack')
--b0.messagePack.set_string('binary')
--b0.messagePack.set_string('string')

function b0RemoteApi(nodeName,channelName,inactivityToleranceInSec,setupSubscribersAsynchronously)
    local self={}
    
    if nodeName==nil then nodeName='b0RemoteApi_luaClient' end
    if channelName==nil then channelName='b0RemoteApi' end
    if inactivityToleranceInSec==nil then inactivityToleranceInSec=60 end
    if setupSubscribersAsynchronously==nil then setupSubscribersAsynchronously=false end
    
    local _channelName=channelName
    local _serviceCallTopic=channelName..'SerX'
    local _defaultPublisherTopic=channelName..'SubX'
    local _defaultSubscriberTopic=channelName..'PubX'
    local _nextDefaultSubscriberHandle=2
    local _nextDedicatedPublisherHandle=500
    local _nextDedicatedSubscriberHandle=1000
    local _node=b0.node_new(nodeName)
    math.randomseed(b0.node_hardware_time_usec(_node))
    local _clientId=''
    for i=1,10,1 do
        local r=math.random(62)
        _clientId=_clientId..string.sub('abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789',r,r)
    end
    local _serviceClient=b0.service_client_new_ex(_node,_serviceCallTopic,1,1)
    local _defaultPublisher=b0.publisher_new_ex(_node,_defaultPublisherTopic,1,1)
    local _defaultSubscriber=b0.subscriber_new_ex(_node,_defaultSubscriberTopic,1,1) -- we will poll the socket
    local _allSubscribers={}
    local _allDedicatedPublishers={}
    local _setupSubscribersAsynchronously=setupSubscribersAsynchronously
    local _pongReceived=false

    function self._pingCallback(msg)
        _pongReceived=true
    end
        
    function self.destroy()
        _pongReceived=false
        self._handleFunction('Ping',{0},self.simxDefaultSubscriber(self._pingCallback))
        while not _pongReceived do
            self.simxSpinOnce()
        end
        self._handleFunction('DisconnectClient',{_clientId},_serviceCallTopic)
        for key,value in pairs(_allSubscribers) do 
            if value.handle~=_defaultSubscriber then
                b0.subscriber_delete(value.handle)
            end
        end
        for key,value in pairs(_allDedicatedPublishers) do
            b0.publisher_delete(value)
        end
        b0.node_delete(_node)
    end
    
    function self._handleFunction(funcName,reqArgs,topic)
        if topic==_serviceCallTopic then
            local packedData=b0.messagePack.pack({{funcName,_clientId,topic,0},reqArgs})
            local repl=b0.messagePack.unpack(b0.service_client_call(_serviceClient,packedData))
            return repl
        elseif topic==_defaultPublisherTopic then
            local packedData=b0.messagePack.pack({{funcName,_clientId,topic,1},reqArgs})
            b0.publisher_publish(_defaultPublisher,packedData)
        elseif _allSubscribers[topic] then
            if _allSubscribers[topic].handle==_defaultSubscriber then
                local packedData=b0.messagePack.pack({{funcName,_clientId,topic,2},reqArgs})
                if _setupSubscribersAsynchronously then
                    b0.publisher_publisher(_defaultPublisher,packedData)
                else
                    b0.service_client_call(_serviceClient,packedData)
                end
            else
                local packedData=b0.messagePack.pack({{funcName,_clientId,topic,4},reqArgs})
                if _setupSubscribersAsynchronously then
                    b0.publisher_publish(_defaultPublisher,packedData)
                else
                    b0.service_client_call(_serviceClient,packedData)
                end
            end
        elseif _allDedicatedPublishers[topic] then
            local packedData=b0.messagePack.pack({{funcName,_clientId,topic,3},reqArgs})
            b0.publisher_publish(_allDedicatedPublishers[topic],packedData)
        else
            print('B0 Remote API error: invalid topic')
        end
    end
    
    function self.simxDefaultPublisher()
        return _defaultPublisherTopic
    end

    function self.simxCreatePublisher(dropMessages)
        if dropMessages==nil then dropMessages=false end
        local topic=_channelName..'Sub'..tostring(_nextDedicatedPublisherHandle).._clientId
        _nextDedicatedPublisherHandle=_nextDedicatedPublisherHandle+1
        local pub=b0.publisher_new_ex(_node,topic,0,1)
        b0.publisher_init(pub)
        _allDedicatedPublishers[topic]=pub
        self._handleFunction('createSubscriber',{topic,dropMessages},_serviceCallTopic)
        return topic
    end

    function self.simxDefaultSubscriber(cb,publishInterval)
        if publishInterval==nil then publishInterval=1 end
        local topic=_channelName..'Pub'..tostring(_nextDefaultSubscriberHandle).._clientId
        _nextDefaultSubscriberHandle=_nextDefaultSubscriberHandle+1
        _allSubscribers[topic]={}
        _allSubscribers[topic].handle=_defaultSubscriber
        _allSubscribers[topic].cb=cb
        _allSubscribers[topic].dropMessages=false
        local channel=_serviceCallTopic
        if _setupSubscribersAsynchronously then
            channel=_defaultPublisherTopic
        end
        self._handleFunction('setDefaultPublisherPubInterval',{topic,publishInterval},channel)
        return topic
    end
        
    function self.simxCreateSubscriber(cb,publishInterval,dropMessages)
        if publishInterval==nil then publishInterval=1 end
        if dropMessages==nil then dropMessages=false end
        local topic=_channelName..'Pub'..tostring(_nextDedicatedSubscriberHandle).._clientId
        _nextDedicatedSubscriberHandle=_nextDedicatedSubscriberHandle+1
        local subb=b0.subscriber_new_ex(_node,topic,0,1)
        b0.subscriber_init(subb)
        _allSubscribers[topic]={}
        _allSubscribers[topic].handle=subb
        _allSubscribers[topic].cb=cb
        _allSubscribers[topic].dropMessages=dropMessages
        local channel=_serviceCallTopic
        if _setupSubscribersAsynchronously then
            channel=_defaultPublisherTopic
        end
        self._handleFunction('createPublisher',{topic,publishInterval},channel)
        return topic
    end
  
    function self.simxServiceCall()
        return _serviceCallTopic
    end

    function self._handleReceivedMessage(msg)
        msg=b0.messagePack.unpack(msg)
        if _allSubscribers[msg[1]] then
            local cbMsg=msg[2]
            _allSubscribers[msg[1]].cb(cbMsg)
        end
    end
        
    function self.simxSpinOnce()
        local defaultSubscriberAlreadyProcessed=false
        for key,value in pairs(_allSubscribers) do
            local readData=nil
            if (value.handle~=_defaultSubscriber) or (not defaultSubscriberAlreadyProcessed) then
                defaultSubscriberAlreadyProcessed=defaultSubscriberAlreadyProcessed or (value.handle==_defaultSubscriber)
                while b0.subscriber_poll(value.handle,0)>0 do
                    readData=b0.subscriber_read(value.handle)
                    if not value.dropMessages then
                        self._handleReceivedMessage(readData)
                    end
                end
                if value.dropMessages and readData then
                    self._handleReceivedMessage(readData)
                end
            end
        end
    end
                    
    function self.simxSpin()
        while true do
            self.simxSpinOnce()
        end
    end

    function self.simxSynchronous(enable)
        local reqArgs = {enable}
        local funcName = 'Synchronous'
        self._handleFunction(funcName,reqArgs,_serviceCallTopic)
    end
        
    function self.simxSynchronousTrigger()
        local reqArgs = {0}
        local funcName = 'SynchronousTrigger'
        self._handleFunction(funcName,reqArgs,_defaultPublisherTopic)
    end
        
    function self.simxGetSimulationStepDone(topic)
        if _allSubscribers[topic] then
            local reqArgs = {0}
            local funcName = 'GetSimulationStepDone'
            self._handleFunction(funcName,reqArgs,topic)
        else
            print('B0 Remote API error: invalid topic')
        end
    end
        
    function self.simxGetSimulationStepStarted(topic)
        if _allSubscribers[topic] then
            local reqArgs = {0}
            local funcName = 'GetSimulationStepStarted'
            self._handleFunction(funcName,reqArgs,topic)
        else
            print('B0 Remote API error: invalid topic')
        end
    end
    
    function self.simxGetTimeInMs()
        return b0.node_hardware_time_usec(_node)/1000    
    end
    
    function self.simxSleep(durationInMs)
        local st=self.simxGetTimeInMs()
        while self.simxGetTimeInMs()-st<durationInMs do end
    end
    
    print('\n  Running B0 Remote API client with channel name ['..channelName..']')
    print('  make sure that: 1) the B0 resolver is running')
    print('                  2) V-REP is running the B0 Remote API server with the same channel name')
    print('  Initializing...\n')
    b0.node_init(_node)
    
    self._handleFunction('inactivityTolerance',{inactivityToleranceInSec},_serviceCallTopic)
    print('\n  Connected!\n')
    
    -- ------------------------------
    -- Add your custom function here:
    -- ------------------------------
        
    function self.simxGetObjectHandle(objectName,topic)
        local reqArgs = {objectName}
        local funcName = 'GetObjectHandle'
        return self._handleFunction(funcName,reqArgs,topic)
    end

    function self.simxAddStatusbarMessage(txt,topic)
        local reqArgs = {txt}
        local funcName = 'AddStatusbarMessage'
        return self._handleFunction(funcName,reqArgs,topic)
    end
        
    function self.simxStartSimulation(topic)
        local reqArgs = {}
        local funcName = 'StartSimulation'
        return self._handleFunction(funcName,reqArgs,topic)
    end
        
    function self.simxStopSimulation(topic)
        local reqArgs = {}
        local funcName = 'StopSimulation'
        return self._handleFunction(funcName,reqArgs,topic)
    end

    function self.simxGetVisionSensorImage(objectHandle,greyscale,topic)
        local reqArgs = {objectHandle,greyscale}
        local funcName = 'GetVisionSensorImage'
        return self._handleFunction(funcName,reqArgs,topic)
    end

    function self.simxSetVisionSensorImage(objectHandle,greyscale,img,topic)
        local reqArgs = {objectHandle,greyscale,img}
        local funcName = 'SetVisionSensorImage'
        return self._handleFunction(funcName,reqArgs,topic)
    end

--[==[        
        
        
        
    function self.simxAuxiliaryConsoleClose(self,consoleHandle,topic)
        local reqArgs = {consoleHandle}
        local funcName = 'AuxiliaryConsoleClose'
        return self._handleFunction(funcName,reqArgs,topic)
        
    function self.simxAuxiliaryConsolePrint(self,consoleHandle,text,topic)
        local reqArgs = {consoleHandle,text}
        local funcName = 'AuxiliaryConsolePrint'
        return self._handleFunction(funcName,reqArgs,topic)
        
    function self.simxAuxiliaryConsoleOpen(self,title,maxLines,mode,position,size,textColor,backgroundColor,topic)
        local reqArgs = {title,maxLines,mode,position,size,textColor,backgroundColor}
        local funcName = 'AuxiliaryConsoleOpen'
        return self._handleFunction(funcName,reqArgs,topic)
        
    function self.simxAuxiliaryConsoleShow(self,consoleHandle,showState,topic)
        local reqArgs = {consoleHandle,showState}
        local funcName = 'AuxiliaryConsoleShow'
        return self._handleFunction(funcName,reqArgs,topic)
        

    function self.simxGetObjectPosition(self,objectHandle,refObjectHandle,topic)
        local reqArgs = {objectHandle,refObjectHandle}
        local funcName = 'GetObjectPosition'
        return self._handleFunction(funcName,reqArgs,topic)

    function self.simxAddDrawingObject_points(self,size,color,coords,topic)
        local reqArgs = {size,color,coords}
        local funcName = 'AddDrawingObject_points'
        return self._handleFunction(funcName,reqArgs,topic)

    function self.simxAddDrawingObject_spheres(self,size,color,coords,topic)
        local reqArgs = {size,color,coords}
        local funcName = 'AddDrawingObject_spheres'
        return self._handleFunction(funcName,reqArgs,topic)

    function self.simxAddDrawingObject_cubes(self,size,color,coords,topic)
        local reqArgs = {size,color,coords}
        local funcName = 'AddDrawingObject_cubes'
        return self._handleFunction(funcName,reqArgs,topic)

    function self.simxAddDrawingObject_lines(self,lineSize,color,segments,topic)
        local reqArgs = {lineSize,color,segments}
        local funcName = 'AddDrawingObject_lines'
        return self._handleFunction(funcName,reqArgs,topic)

    function self.simxAddDrawingObject_triangles(self,color,triangles,topic)
        local reqArgs = {color,triangles}
        local funcName = 'AddDrawingObject_triangles'
        return self._handleFunction(funcName,reqArgs,topic)

    function self.simxRemoveDrawingObject(self,handle,topic)
        local reqArgs = {handle}
        local funcName = 'RemoveDrawingObject'
        return self._handleFunction(funcName,reqArgs,topic)
        
    function self.simxCallScriptFunction(self,funcAtObjName,scriptType,arg1,arg2,arg3,arg4,topic)
        local reqArgs = {funcAtObjName,scriptType,arg1,arg2,arg3,arg4}
        local funcName = 'CallScriptFunction'
        return self._handleFunction(funcName,reqArgs,topic)
        
    function self.simxCheckCollision(self,entity1,entity2,topic)
        local reqArgs = {entity1,entity2}
        local funcName = 'CheckCollision'
        return self._handleFunction(funcName,reqArgs,topic)
        
    function self.simxGetCollisionHandle(self,name,topic)
        local reqArgs = {name}
        local funcName = 'GetCollisionHandle'
        return self._handleFunction(funcName,reqArgs,topic)
        
    function self.simxReadCollision(self,handle,topic)
        local reqArgs = {handle}
        local funcName = 'ReadCollision'
        return self._handleFunction(funcName,reqArgs,topic)
        
    function self.simxCheckDistance(self,entity1,entity2,threshold,topic)
        local reqArgs = {entity1,entity2,threshold}
        local funcName = 'CheckDistance'
        return self._handleFunction(funcName,reqArgs,topic)
        
    function self.simxGetDistanceHandle(self,name,topic)
        local reqArgs = {name}
        local funcName = 'GetDistanceHandle'
        return self._handleFunction(funcName,reqArgs,topic)
        
    function self.simxReadDistance(self,handle,topic)
        local reqArgs = {handle}
        local funcName = 'ReadDistance'
        return self._handleFunction(funcName,reqArgs,topic)
        
    function self.simxCheckProximitySensor(self,sensor,entity,topic)
        local reqArgs = {sensor,entity}
        local funcName = 'CheckProximitySensor'
        return self._handleFunction(funcName,reqArgs,topic)
        
    function self.simxReadProximitySensor(self,handle,topic)
        local reqArgs = {handle}
        local funcName = 'ReadProximitySensor'
        return self._handleFunction(funcName,reqArgs,topic)
        
    function self.simxCheckVisionSensor(self,sensor,entity,topic)
        local reqArgs = {sensor,entity}
        local funcName = 'CheckVisionSensor'
        return self._handleFunction(funcName,reqArgs,topic)
        
    function self.simxReadVisionSensor(self,handle,topic)
        local reqArgs = {handle}
        local funcName = 'ReadVisionSensor'
        return self._handleFunction(funcName,reqArgs,topic)
        
    function self.simxReadForceSensor(self,handle,topic)
        local reqArgs = {handle}
        local funcName = 'ReadForceSensor'
        return self._handleFunction(funcName,reqArgs,topic)
        
    function self.simxBreakForceSensor(self,handle,topic)
        local reqArgs = {handle}
        local funcName = 'BreakForceSensor'
        return self._handleFunction(funcName,reqArgs,topic)
        
    function self.simxClearFloatSignal(self,sig,topic)
        local reqArgs = {sig}
        local funcName = 'ClearFloatSignal'
        return self._handleFunction(funcName,reqArgs,topic)
        
    function self.simxClearIntegerSignal(self,sig,topic)
        local reqArgs = {sig}
        local funcName = 'ClearIntegerSignal'
        return self._handleFunction(funcName,reqArgs,topic)
        
    function self.simxClearStringSignal(self,sig,topic)
        local reqArgs = {sig}
        local funcName = 'ClearStringSignal'
        return self._handleFunction(funcName,reqArgs,topic)
        
    function self.simxSetFloatSignal(self,sig,val,topic)
        local reqArgs = {sig,val}
        local funcName = 'SetFloatSignal'
        return self._handleFunction(funcName,reqArgs,topic)
        
    function self.simxSetIntegerSignal(self,sig,val,topic)
        local reqArgs = {sig,val}
        local funcName = 'SetIntegerSignal'
        return self._handleFunction(funcName,reqArgs,topic)
        
    function self.simxSetStringSignal(self,sig,val,topic)
        local reqArgs = {sig,val}
        local funcName = 'SetStringSignal'
        return self._handleFunction(funcName,reqArgs,topic)
        
    function self.simxGetFloatSignal(self,sig,topic)
        local reqArgs = {sig}
        local funcName = 'GetFloatSignal'
        return self._handleFunction(funcName,reqArgs,topic)
        
    function self.simxGetIntegerSignal(self,sig,topic)
        local reqArgs = {sig}
        local funcName = 'GetIntegerSignal'
        return self._handleFunction(funcName,reqArgs,topic)
        
    function self.simxGetStringSignal(self,sig,topic)
        local reqArgs = {sig}
        local funcName = 'GetStringSignal'
        return self._handleFunction(funcName,reqArgs,topic)
--]==]    
    
    return self
end

File: python/b0RemoteApi.py
# -------------------------------------------------------
# Add your custom functions at the bottom of the file
# and the server counterpart to lua/b0RemoteApiServer.lua
# -------------------------------------------------------

import b0
import msgpack
import random
import string
import time

class RemoteApiClient:
    def __init__(self,nodeName='b0RemoteApi_pythonClient',channelName='b0RemoteApi',inactivityToleranceInSec=60,setupSubscribersAsynchronously=False):
        self._channelName=channelName
        self._serviceCallTopic=channelName+'SerX'
        self._defaultPublisherTopic=channelName+'SubX'
        self._defaultSubscriberTopic=channelName+'PubX'
        self._nextDefaultSubscriberHandle=2
        self._nextDedicatedPublisherHandle=500
        self._nextDedicatedSubscriberHandle=1000
        self._node=b0.Node(nodeName)
        self._clientId=''.join(random.choice(string.ascii_uppercase+string.ascii_lowercase+string.digits) for _ in range(10))
        self._serviceClient=b0.ServiceClient(self._node,self._serviceCallTopic)
        self._defaultPublisher=b0.Publisher(self._node,self._defaultPublisherTopic)
        self._defaultSubscriber=b0.Subscriber(self._node,self._defaultSubscriberTopic,None) # we will poll the socket
        print('\n  Running B0 Remote API client with channel name ['+channelName+']')
        print('  make sure that: 1) the B0 resolver is running')
        print('                  2) V-REP is running the B0 Remote API server with the same channel name')
        print('  Initializing...\n')
        self._node.init()
        self._handleFunction('inactivityTolerance',[inactivityToleranceInSec],self._serviceCallTopic)
        print('\n  Connected!\n')
        self._allSubscribers={}
        self._allDedicatedPublishers={}
        self._setupSubscribersAsynchronously=setupSubscribersAsynchronously
  
    def __enter__(self):
        return self
    
    def __exit__(self,*err):
        self._pongReceived=False
        self._handleFunction('Ping',[0],self.simxDefaultSubscriber(self._pingCallback))
        while not self._pongReceived:
            self.simxSpinOnce();
        self._handleFunction('DisconnectClient',[self._clientId],self._serviceCallTopic)
        for key, value in self._allSubscribers.items():
            if value['handle']!=self._defaultSubscriber:
                value['handle'].cleanup()
        for key, value in self._allDedicatedPublishers.items():
            value.cleanup()
        self._node.cleanup()
        
    def _pingCallback(self,msg):
        self._pongReceived=True
        
    def _handleReceivedMessage(self,msg):
        msg=msgpack.unpackb(msg)
        msg[0]=msg[0].decode('ascii')
        if msg[0] in self._allSubscribers:
            cbMsg=msg[1]
            if len(cbMsg)==1:
                cbMsg.append(None)
            self._allSubscribers[msg[0]]['cb'](cbMsg)
            
    def _handleFunction(self,funcName,reqArgs,topic):
        if topic==self._serviceCallTopic:
            packedData=msgpack.packb([[funcName,self._clientId,topic,0],reqArgs])
            rep = msgpack.unpackb(self._serviceClient.call(packedData))
            if len(rep)==1:
                rep.append(None)
            return rep
        elif topic==self._defaultPublisherTopic:
            packedData=msgpack.packb([[funcName,self._clientId,topic,1],reqArgs])
            self._defaultPublisher.publish(packedData)
        elif topic in self._allSubscribers:
            if self._allSubscribers[topic]['handle']==self._defaultSubscriber:
                packedData=msgpack.packb([[funcName,self._clientId,topic,2],reqArgs])
                if self._setupSubscribersAsynchronously:
                    self._defaultPublisher.publish(packedData)
                else:
                    self._serviceClient.call(packedData)
            else:
                packedData=msgpack.packb([[funcName,self._clientId,topic,4],reqArgs])
                if self._setupSubscribersAsynchronously:
                    self._defaultPublisher.publish(packedData)
                else:
                    self._serviceClient.call(packedData)
        elif topic in self._allDedicatedPublishers:
            packedData=msgpack.packb([[funcName,self._clientId,topic,3],reqArgs])
            self._allDedicatedPublishers[topic].publish(packedData)
        else:
            print('B0 Remote API error: invalid topic')
        
    def simxDefaultPublisher(self):
        return self._defaultPublisherTopic

    def simxCreatePublisher(self,dropMessages=False):
        topic=self._channelName+'Sub'+str(self._nextDedicatedPublisherHandle)+self._clientId
        self._nextDedicatedPublisherHandle=self._nextDedicatedPublisherHandle+1
        pub=b0.Publisher(self._node,topic,0,1)
        pub.init()
        self._allDedicatedPublishers[topic]=pub
        self._handleFunction('createSubscriber',[topic,dropMessages],self._serviceCallTopic)
        return topic

    def simxDefaultSubscriber(self,cb,publishInterval=1):
        topic=self._channelName+'Pub'+str(self._nextDefaultSubscriberHandle)+self._clientId
        self._nextDefaultSubscriberHandle=self._nextDefaultSubscriberHandle+1
        self._allSubscribers[topic]={}
        self._allSubscribers[topic]['handle']=self._defaultSubscriber
        self._allSubscribers[topic]['cb']=cb
        self._allSubscribers[topic]['dropMessages']=False
        channel=self._serviceCallTopic
        if self._setupSubscribersAsynchronously:
            channel=self._defaultPublisherTopic
        self._handleFunction('setDefaultPublisherPubInterval',[topic,publishInterval],channel)
        return topic
        
    def simxCreateSubscriber(self,cb,publishInterval=1,dropMessages=False):
        topic=self._channelName+'Pub'+str(self._nextDedicatedSubscriberHandle)+self._clientId
        self._nextDedicatedSubscriberHandle=self._nextDedicatedSubscriberHandle+1
        sub=b0.Subscriber(self._node,topic,None,0,1)
        sub.init()
        self._allSubscribers[topic]={}
        self._allSubscribers[topic]['handle']=sub
        self._allSubscribers[topic]['cb']=cb
        self._allSubscribers[topic]['dropMessages']=dropMessages
        channel=self._serviceCallTopic
        if self._setupSubscribersAsynchronously:
            channel=self._defaultPublisherTopic
        self._handleFunction('createPublisher',[topic,publishInterval],channel)
        return topic
  
    def simxServiceCall(self):
        return self._serviceCallTopic
        
    def simxSpin(self):
        while True:
            self.simxSpinOnce()
        
    def simxSpinOnce(self):
        defaultSubscriberAlreadyProcessed=False
        for key, value in self._allSubscribers.items():
            readData=None
            if (value['handle']!=self._defaultSubscriber) or (not defaultSubscriberAlreadyProcessed):
                defaultSubscriberAlreadyProcessed=defaultSubscriberAlreadyProcessed or (value['handle']==self._defaultSubscriber)
                while value['handle'].poll(0):
                    readData=value['handle'].read()
                    if not value['dropMessages']:
                        self._handleReceivedMessage(readData)
                if value['dropMessages'] and (readData is not None):
                    self._handleReceivedMessage(readData)
                    
    def simxGetTimeInMs(self):
        return self._node.hardware_time_usec()/1000;    

    def simxSleep(self,durationInMs):
        time.sleep(durationInMs)
        
    def simxSynchronous(self,enable):
        reqArgs = [enable]
        funcName = 'Synchronous'
        self._handleFunction(funcName,reqArgs,self._serviceCallTopic)
        
    def simxSynchronousTrigger(self):
        reqArgs = [0]
        funcName = 'SynchronousTrigger'
        self._handleFunction(funcName,reqArgs,self._defaultPublisherTopic)
        
    def simxGetSimulationStepDone(self,topic):
        if topic in self._allSubscribers:
            reqArgs = [0]
            funcName = 'GetSimulationStepDone'
            self._handleFunction(funcName,reqArgs,topic)
        else:
            print('B0 Remote API error: invalid topic')
        
    def simxGetSimulationStepStarted(self,topic):
        if topic in self._allSubscribers:
            reqArgs = [0]
            funcName = 'GetSimulationStepStarted'
            self._handleFunction(funcName,reqArgs,topic)
        else:
            print('B0 Remote API error: invalid topic')
    
    def simxAuxiliaryConsoleClose(self,consoleHandle,topic):
        reqArgs = [consoleHandle]
        funcName = 'AuxiliaryConsoleClose'
        return self._handleFunction(funcName,reqArgs,topic)
        
    def simxAuxiliaryConsolePrint(self,consoleHandle,text,topic):
        reqArgs = [consoleHandle,text]
        funcName = 'AuxiliaryConsolePrint'
        return self._handleFunction(funcName,reqArgs,topic)
        
    def simxAuxiliaryConsoleOpen(self,title,maxLines,mode,position,size,textColor,backgroundColor,topic):
        reqArgs = [title,maxLines,mode,position,size,textColor,backgroundColor]
        funcName = 'AuxiliaryConsoleOpen'
        return self._handleFunction(funcName,reqArgs,topic)
        
    def simxAuxiliaryConsoleShow(self,consoleHandle,showState,topic):
        reqArgs = [consoleHandle,showState]
        funcName = 'AuxiliaryConsoleShow'
        return self._handleFunction(funcName,reqArgs,topic)
        
    def simxGetObjectHandle(self,objectName,topic):
        reqArgs = [objectName]
        funcName = 'GetObjectHandle'
        return self._handleFunction(funcName,reqArgs,topic)

    def simxAddStatusbarMessage(self,txt,topic):
        reqArgs = [txt]
        funcName = 'AddStatusbarMessage'
        return self._handleFunction(funcName,reqArgs,topic)

    def simxGetObjectPosition(self,objectHandle,refObjectHandle,topic):
        reqArgs = [objectHandle,refObjectHandle]
        funcName = 'GetObjectPosition'
        return self._handleFunction(funcName,reqArgs,topic)

    def simxStartSimulation(self,topic):
        reqArgs = []
        funcName = 'StartSimulation'
        return self._handleFunction(funcName,reqArgs,topic)
        
    def simxStopSimulation(self,topic):
        reqArgs = []
        funcName = 'StopSimulation'
        return self._handleFunction(funcName,reqArgs,topic)

    def simxGetVisionSensorImage(self,objectHandle,greyscale,topic):
        reqArgs = [objectHandle,greyscale]
        funcName = 'GetVisionSensorImage'
        return self._handleFunction(funcName,reqArgs,topic)

    def simxSetVisionSensorImage(self,objectHandle,greyscale,img,topic):
        reqArgs = [objectHandle,greyscale,img]
        funcName = 'SetVisionSensorImage'
        return self._handleFunction(funcName,reqArgs,topic)

    def simxAddDrawingObject_points(self,size,color,coords,topic):
        reqArgs = [size,color,coords]
        funcName = 'AddDrawingObject_points'
        return self._handleFunction(funcName,reqArgs,topic)

    def simxAddDrawingObject_spheres(self,size,color,coords,topic):
        reqArgs = [size,color,coords]
        funcName = 'AddDrawingObject_spheres'
        return self._handleFunction(funcName,reqArgs,topic)

    def simxAddDrawingObject_cubes(self,size,color,coords,topic):
        reqArgs = [size,color,coords]
        funcName = 'AddDrawingObject_cubes'
        return self._handleFunction(funcName,reqArgs,topic)

    def simxAddDrawingObject_segments(self,lineSize,color,segments,topic):
        reqArgs = [lineSize,color,segments]
        funcName = 'AddDrawingObject_segments'
        return self._handleFunction(funcName,reqArgs,topic)

    def simxAddDrawingObject_triangles(self,color,triangles,topic):
        reqArgs = [color,triangles]
        funcName = 'AddDrawingObject_triangles'
        return self._handleFunction(funcName,reqArgs,topic)

    def simxRemoveDrawingObject(self,handle,topic):
        reqArgs = [handle]
        funcName = 'RemoveDrawingObject'
        return self._handleFunction(funcName,reqArgs,topic)
        
    def simxCallScriptFunction(self,funcAtObjName,scriptType,arg1,arg2,arg3,arg4,topic):
        reqArgs = [funcAtObjName,scriptType,arg1,arg2,arg3,arg4]
        funcName = 'CallScriptFunction'
        return self._handleFunction(funcName,reqArgs,topic)
        
    def simxCheckCollision(self,entity1,entity2,topic):
        reqArgs = [entity1,entity2]
        funcName = 'CheckCollision'
        return self._handleFunction(funcName,reqArgs,topic)
        
    def simxGetCollisionHandle(self,name,topic):
        reqArgs = [name]
        funcName = 'GetCollisionHandle'
        return self._handleFunction(funcName,reqArgs,topic)
        
    def simxReadCollision(self,handle,topic):
        reqArgs = [handle]
        funcName = 'ReadCollision'
        return self._handleFunction(funcName,reqArgs,topic)
        
    def simxCheckDistance(self,entity1,entity2,threshold,topic):
        reqArgs = [entity1,entity2,threshold]
        funcName = 'CheckDistance'
        return self._handleFunction(funcName,reqArgs,topic)
        
    def simxGetDistanceHandle(self,name,topic):
        reqArgs = [name]
        funcName = 'GetDistanceHandle'
        return self._handleFunction(funcName,reqArgs,topic)
        
    def simxReadDistance(self,handle,topic):
        reqArgs = [handle]
        funcName = 'ReadDistance'
        return self._handleFunction(funcName,reqArgs,topic)
        
    def simxCheckProximitySensor(self,sensor,entity,topic):
        reqArgs = [sensor,entity]
        funcName = 'CheckProximitySensor'
        return self._handleFunction(funcName,reqArgs,topic)
        
    def simxReadProximitySensor(self,handle,topic):
        reqArgs = [handle]
        funcName = 'ReadProximitySensor'
        return self._handleFunction(funcName,reqArgs,topic)
        
    def simxCheckVisionSensor(self,sensor,entity,topic):
        reqArgs = [sensor,entity]
        funcName = 'CheckVisionSensor'
        return self._handleFunction(funcName,reqArgs,topic)
        
    def simxReadVisionSensor(self,handle,topic):
        reqArgs = [handle]
        funcName = 'ReadVisionSensor'
        return self._handleFunction(funcName,reqArgs,topic)
        
    def simxReadForceSensor(self,handle,topic):
        reqArgs = [handle]
        funcName = 'ReadForceSensor'
        return self._handleFunction(funcName,reqArgs,topic)
        
    def simxBreakForceSensor(self,handle,topic):
        reqArgs = [handle]
        funcName = 'BreakForceSensor'
        return self._handleFunction(funcName,reqArgs,topic)
        
    def simxClearFloatSignal(self,sig,topic):
        reqArgs = [sig]
        funcName = 'ClearFloatSignal'
        return self._handleFunction(funcName,reqArgs,topic)
        
    def simxClearIntegerSignal(self,sig,topic):
        reqArgs = [sig]
        funcName = 'ClearIntegerSignal'
        return self._handleFunction(funcName,reqArgs,topic)
        
    def simxClearStringSignal(self,sig,topic):
        reqArgs = [sig]
        funcName = 'ClearStringSignal'
        return self._handleFunction(funcName,reqArgs,topic)
        
    def simxSetFloatSignal(self,sig,val,topic):
        reqArgs = [sig,val]
        funcName = 'SetFloatSignal'
        return self._handleFunction(funcName,reqArgs,topic)
        
    def simxSetIntegerSignal(self,sig,val,topic):
        reqArgs = [sig,val]
        funcName = 'SetIntegerSignal'
        return self._handleFunction(funcName,reqArgs,topic)
        
    def simxSetStringSignal(self,sig,val,topic):
        reqArgs = [sig,val]
        funcName = 'SetStringSignal'
        return self._handleFunction(funcName,reqArgs,topic)
        
    def simxGetFloatSignal(self,sig,topic):
        reqArgs = [sig]
        funcName = 'GetFloatSignal'
        return self._handleFunction(funcName,reqArgs,topic)
        
    def simxGetIntegerSignal(self,sig,topic):
        reqArgs = [sig]
        funcName = 'GetIntegerSignal'
        return self._handleFunction(funcName,reqArgs,topic)
        
    def simxGetStringSignal(self,sig,topic):
        reqArgs = [sig]
        funcName = 'GetStringSignal'
        return self._handleFunction(funcName,reqArgs,topic)

    # -------------------------------
    # Add your custom functions here:
    # -------------------------------
        